src/chainstate/stacks/db/transactions.rs

```diff
@@ -8417,4 +8530,286 @@ pub mod test {
             assert!(false)
         };
     }
+
+    #[test]
+    fn test_checkerrors_at_runtime() {
+        let runtime_checkerror_trait = "
+            (define-trait foo
+                (
+                    (lolwut () (response bool uint))
+                )
+            )
+            "
+        .to_string();
+
+        let runtime_checkerror_impl = "
+            (impl-trait .foo.foo)
+
+            (define-public (lolwut)
+                (ok true)
+            )
+            "
+        .to_string();
+
+        let runtime_checkerror = "
+            (use-trait trait .foo.foo)
+
+            (define-data-var mutex bool true)
+
+            (define-public (flip)
+              (ok (var-set mutex (not (var-get mutex))))
+            )
+
+            ;; triggers checkerror at runtime because <trait> gets coerced
+            ;; into a principal when `internal` is called.
+            (define-public (test (ref <trait>))
+                (ok (internal (if (var-get mutex)
+                    (some ref)
+                    none
+                )))
+            )
+
+            (define-private (internal (ref (optional <trait>))) true)
+            "
+        .to_string();
+
+        let privk = StacksPrivateKey::from_hex(
+            "6d430bb91222408e7706c9001cfaeb91b08c2be6d5ac95779ab52c6b431950e001",
+        )
+        .unwrap();
+        let auth = TransactionAuth::from_p2pkh(&privk).unwrap();
+        let addr = auth.origin().address_testnet();
+
+        let balances = vec![(addr.clone(), 1000000000)];
+
+        let mut chainstate = instantiate_chainstate_with_balances(
+            false,
+            0x80000000,
+            "test_checkerrors_at_runtime",
+            balances,
+        );
+
+        let mut tx_runtime_checkerror_trait = StacksTransaction::new(
+            TransactionVersion::Testnet,
+            auth.clone(),
+            TransactionPayload::new_smart_contract(
+                &"foo".to_string(),
+                &runtime_checkerror_trait.to_string(),
+                None,
+            )
+            .unwrap(),
+        );
+
+        tx_runtime_checkerror_trait.post_condition_mode = TransactionPostConditionMode::Allow;
+        tx_runtime_checkerror_trait.chain_id = 0x80000000;
+        tx_runtime_checkerror_trait.set_tx_fee(0);
+        tx_runtime_checkerror_trait.set_origin_nonce(0);
+
+        let mut signer = StacksTransactionSigner::new(&tx_runtime_checkerror_trait);
+        signer.sign_origin(&privk).unwrap();
+
+        let signed_runtime_checkerror_trait_tx = signer.get_tx().unwrap();
+
+        let mut tx_runtime_checkerror_impl = StacksTransaction::new(
+            TransactionVersion::Testnet,
+            auth.clone(),
+            TransactionPayload::new_smart_contract(
+                &"foo-impl".to_string(),
+                &runtime_checkerror_trait.to_string(),
```

Comment:
Is this supposed to be `runtime_checkerror_impl`?

Reply:
Fixed